### PR TITLE
proxy_get for Riak EE 1.4

### PR DESCRIPTION
**Pull request by bookshelfdave**

### CODE REVIEW

**Comment by bookshelfdave** [comment]
See also:
https://github.com/basho/riak_ee/pull/134
https://github.com/basho/riak_test/pull/273
https://github.com/basho/riak_cs/pull/516

File: src/riak_repl_pb_get.erl [hunk 1]
```diff
@@ -76,31 +87,53 @@ process(#rpbreplgetreq{bucket=B, key=K, r=R0, pr=PR0, notfound_ok=NFOk,
...
+                    false ->
+                        notconnected
+                end,
+            lager:info("Result12 = ~p", [ Result12 ]),
+            lager:info("Result13 = ~p", [ Result13 ]),
+            Result =
+                case {Result12, Result13} of
+                    {notconnected, Value} -> Value;
+                    {Value, notconnected} -> Value;
+                    _ -> notconnected
```
Comment (Vagabond) [inline src/riak_repl_pb_get.erl:121]: What if you're connected to a 1.3 cluster and a 1.2 cluster and both return cluster name responses?

File: src/riak_repl_ring.erl [hunk 2]
```diff
@@ -293,6 +296,17 @@ get_clusters(Ring) ->
             []
     end.
 
+
+% TODO: comments
```
Comment (Vagabond) [inline src/riak_repl_ring.erl]: TODO?

File: src/riak_repl_pb_get.erl [hunk 3]
```diff
@@ -76,31 +87,53 @@ process(#rpbreplgetreq{bucket=B, key=K, r=R0, pr=PR0, notfound_ok=NFOk,
...
-                    lager:debug("not connected to cluster ~p", [CName]),
+            lager:info("CName = ~p", [ CName ]),
+            Modes = State#state.repl_modes,
+            Repl12Enabled = riak_repl_util:mode_12_enabled(Modes),
+            Repl13Enabled = riak_repl_util:mode_13_enabled(Modes),
+            Result12 =
+                case Repl12Enabled of
+                    true ->
+                        CNames12 = get_client_cluster_names_12(),
+                        lager:info("CNames12 = ~p", [ CNames12 ]),
```
Comment (bookshelfdave) [inline src/riak_repl_pb_get.erl]: a lot of these lager:info's should be debugs, I'll change them along during my next commit

File: src/riak_repl_ring_handler.erl [hunk 4]
```diff
@@ -179,3 +181,20 @@ rt_update_events(Ring) ->
...
+    _PGEnabled = app_helper:get_env(riak_repl, pgenabled, false),
+
+    _RC = case riak_repl_ring:get_repl_config(Ring) of
+        undefined ->
+            riak_repl_ring:initial_config();
+        R ->
+            R
+    end,
+
+    % todo
```
Comment (Vagabond) [inline src/riak_repl_ring_handler.erl]: TODO?

File: src/riak_repl_ring.erl (hunk 2 above)
Comment (bookshelfdave) [inline src/riak_repl_ring.erl]: TODONT

File: src/riak_repl_ring_handler.erl [hunk 5]
```diff
@@ -179,3 +181,20 @@ rt_update_events(Ring) ->
     %% the rtenabled environment variable
     riak_repl:install_hook().
 
+pg_update_events(Ring) ->
+    riak_repl2_pg:ensure_pg(riak_repl_ring:pg_enabled(Ring)),
+    
+    %% ensure_rt sets this
+    _PGEnabled = app_helper:get_env(riak_repl, pgenabled, false),
```
Comment (Vagabond) [inline src/riak_repl_ring_handler.erl]: Why all these underscore variables that aren't used?

File: src/riak_repl_ring.erl (hunk 2 above)
Comment (Vagabond) [inline src/riak_repl_ring.erl]: Then remove the TODO.

File: src/riak_repl2_pg.erl [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                [] ->
+                    [];
+                PGPs ->
+                    [{Remote, riak_repl2_pg_block_provider:status(Pid)} || {Remote, Pid} <- PGPs]
+            end,
+            [{proxy_get,[{requester,ReqStats}, {provider, ProStats}]}]
+    end.
+
+enabled() ->
+    {ok, Ring} = riak_core_ring_manager:get_raw_ring(),
```
Comment (Vagabond) [inline src/riak_repl2_pg.erl]: Why do you need the raw ring here?

File: src/riak_repl_ring_handler.erl (hunk 5 above)
Comment (bookshelfdave) [inline src/riak_repl_ring_handler.erl]: These are unneeded, and have been removed along with the TODO below.

File: src/riak_repl2_pg.erl (hunk 6 above)
Comment (bookshelfdave) [inline src/riak_repl2_pg.erl]: I just followed what riak_repl2_rt was doing: https://github.com/basho/riak_repl/blob/master/src/riak_repl2_rt.erl#L47

File: src/riak_repl_pb_get.erl (hunk 1 above)
Comment (bookshelfdave) [inline src/riak_repl_pb_get.erl:121]: How would you feel about:

```
case {Result12, Result13} of
   {notconnected, Value} -> Value;
   {Value, notconnected} -> Value;
   {_, Value} -> Value; %% always default to 1.3
   _ -> notconnected
end
```

File: src/riak_repl2_pg_block_provider.erl [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+-behaviour(gen_server).
+%% API
+-export([start_link/1, connected/6, connect_failed/3]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+    terminate/2, code_change/3, status/1, status/2]).
+
+%% send a message every KEEPALIVE seconds to make sure the service is running on the sink
+-define(KEEPALIVE, 1000).
```
Comment (Vagabond) [inline src/riak_repl2_pg_block_provider.erl:19]: Is this not in milliseconds?

**Comment by Vagabond** [comment]
+1 to merge.

**Comment by bookshelfdave** [comment]
rebased version https://github.com/basho/riak_repl/pull/276

**Pull request closed without merging**